### PR TITLE
Complete overhaul of onboard campaign sorting

### LINKED ISSUE #6863: Feature Card Pulling Campaigns In Wrong Order
#### Current Behavior

Campaigns currently being pulled by staff pick, time and then sponsor. 

*Smiles For Soldiers - .5 hr
# SuperStressFace - .5hr

Note From Shawn - .5hr
Game winning drive (not technically sponsored right now) 2hr
*Power To The Period 5hr

*represents sponsored campaign
#### Desired Behavior

Campaigns should be pulled by staff pick, sponsor, and then time. All sponsored campaigns should appear first but sorted by the amount of time to complete. and then the rest of the staff picks should appear, also sorted by time to complete. Current campaigns should be sorted as:

*smiles for soldiers - .5 hr
*power to the period - 5 hr
superstressface .5 hr
notes from shawn .5 hr
game winning drive - 2hr
#### Why This Matters

It's important for us to prioritize sponsored campaigns for relationship we have with the sponsor.

### PR DESCRIPTION
**Pull request by itsjoekent**
#### What's this PR do?

See the title ^
#### How should this be reviewed?

Rather than clicking through the card to test sorting, I stuck everything in a codepen & wrote two tests at the bottom: http://codepen.io/thedeadlybutter/pen/QEzkJk?editors=0011
This maps out all of the relevant information required to debug & uses prod data, looks perfect to me.

Still test the actual card tho to make sure like it still works.
#### Any background context you want to provide?

Old code was getting hard to follow because we have _so many edge cases_ in this algorithm. 
#### Relevant tickets

Fixes #6863
#### Checklist
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

### CODE REVIEW

File: lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js [hunk 1]
```diff
@@ -25,15 +53,48 @@ export function checkForSponsor(campaign = null) {
 }
 
 /**
- * Sort and rank campaigns by time commitment and if it has a sponsor.
+ * Sort and rank campaigns based on if it has a sponsor.
```
Comment (itsjoekent) [inline lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js]: Whoops, wrong doc block for this function

File: lib/themes/dosomething/paraneue_dosomething/js/components/FeatureCard.js [hunk 2]
```diff
@@ -55,12 +55,8 @@ class FeatureCard extends React.Component {
...
   getCampaigns(staffPickOnly, paginate, page, callback) {
     this.requests.push(this.serverRequest = $.get(`${API_URL}campaigns?${staffPickOnly ? 'staff_pick=true&' : ''}${page != -1 ? 'page=' + page : ''}`, result => {
-
-      // Filter out campaigns that are closed, users have signed up for and then rank remaining campaigns
-      const campaigns = result.data
-        .filter(campaign => campaign.status !== 'closed')
-        .filter(campaign => this.campaignsToExclude.indexOf(parseInt(campaign.id)) == -1)
-        .sort(rankCampaigns);
+      // Get campaigns ordered by relevance & handle exclusions.
+      const campaigns = sortCampaignsByRelevance(result.data).filter(campaign => this.campaignsToExclude.indexOf(parseInt(campaign.id)) == -1);
```
Comment (itsjoekent) [inline lib/themes/dosomething/paraneue_dosomething/js/components/FeatureCard.js:59]: Small optimization, maybe remove exclusions first, then sortByRelevance()

File: lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js [hunk 3]
```diff
@@ -1,4 +1,31 @@
 /**
+ * Checks if the given campaign is still open
+ * @param {object|null} campaign
+ * @return {bool}
+ */
+export function checkIfCampaignIsOpen(campaign = null) {
```
Comment (weerd) [inline lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js]: Maybe rename to `isCampaignOpen`? sounds better when doing a boolean check... "is campaign open? true"

File: lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js [hunk 4]
```diff
@@ -1,4 +1,31 @@
...
+  return campaign.status === 'active';
+}
+
+/**
+ * Checks if the given campaign is a staff pick
+ *
+ * @param {object|null} campaign
+ * @return {bool}
+ */
+export function checkForStaffPick(campaign = null) {
```
Comment (weerd) [inline lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js]: same `isStaffPickCampaign`?

File: lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js (hunk 3 above)
Comment (itsjoekent) [inline lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js]: yeah thats fine, naming isnt my #1 skill lol

Comment (itsjoekent) [inline lib/themes/dosomething/paraneue_dosomething/js/campaign/helpers.js]: oops i linked to issue #1, i never looked at that before

**Comment by DFurnes** [comment]
I know this is outside of the scope of this pull request, but I definitely think we should be looking into making these sorting/filtering updates in the API. Seems crazy to be sending down the whole payload and then doing it client-side.

**Comment by itsjoekent** [comment]
ya that would make things a bit easier!

File: lib/themes/dosomething/paraneue_dosomething/js/components/FeatureCard.js (hunk 2 above)
Comment (weerd) [inline lib/themes/dosomething/paraneue_dosomething/js/components/FeatureCard.js:59]: Yeah, potentially. Not needed for merging but something for later investigation ;)

**Comment by angaither** [comment]
_cough cough_ solr

**Comment by itsjoekent** [comment]
walkinwalkout

**Comment by weerd** [comment]
@DFurnes yeah, initially we felt it would be faster to do this on the client-side without possibly breaking a bunch of other services using the Phoenix API if we made the changes there (especially with no test suite), for something that we're prototyping to see how well it helps with more user activity. If it works out well, and we have the time, definitely should be doing more of that sorting etc on the API side, especially since it has caching in place.

As for Solr... I don't think anyone other than Morgan is willing to deal with that shiz :trollface:

**Comment by angaither** [comment]
it's not that bad. @blisteringherb made a [helpful guide](https://github.com/DoSomething/phoenix/wiki/Solr,-Finder,--campaigns,-Search-Development-and-Troubleshooting)

**Comment by itsjoekent** [comment]
@weerd good to merge?

**Comment by weerd** [comment]
oh yea. If you've tested it thoroughly, give it a whirl 
:sheep: it!

**Comment by itsjoekent** [comment]
the code pen has the two primary tests i used for the sort functionality, i went through every array & it seemed great
